### PR TITLE
chore: Batch writer benchmarks

### PR DESCRIPTION
**Pull request by disq**
https://docs.google.com/spreadsheets/d/1NxKyOtvckW7yCbwLTL4VOu91XPNyd-Q0dYDMxjT-yH8/edit?usp=sharing

Current state (`1000 * N`)
```
GoLand2023.2/tmp/GoLand/___gobench_writers_test_go.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkWriterMemory\E$ -test.run ^$
goos: darwin
goarch: arm64
pkg: github.com/cloudquery/plugin-sdk/v4/writers
BenchmarkWriterMemory
BenchmarkWriterMemory/BatchWriter_batch10k_bytes100M
BenchmarkWriterMemory/BatchWriter_batch10k_bytes100M-10         	     465	   2556097 ns/op	        35.23 bytes/op
BenchmarkWriterMemory/BatchWriter_defaults
BenchmarkWriterMemory/BatchWriter_defaults-10                   	     469	   2595365 ns/op	        25.22 bytes/op
BenchmarkWriterMemory/BatchWriter_wide_batch10k_bytes100M
BenchmarkWriterMemory/BatchWriter_wide_batch10k_bytes100M-10    	       8	 129711115 ns/op	    178917 bytes/op
BenchmarkWriterMemory/BatchWriter_wide_defaults
BenchmarkWriterMemory/BatchWriter_wide_defaults-10              	       8	 151473391 ns/op	     48066 bytes/op
BenchmarkWriterMemory/MixedBatchWriter_batch10k_bytes100M
BenchmarkWriterMemory/MixedBatchWriter_batch10k_bytes100M-10    	     672	   1719326 ns/op	      1394 bytes/op
BenchmarkWriterMemory/MixedBatchWriter_defaults
BenchmarkWriterMemory/MixedBatchWriter_defaults-10              	     585	   1752100 ns/op	      1394 bytes/op
BenchmarkWriterMemory/MixedBatchWriter_wide_batch10k_bytes100M
BenchmarkWriterMemory/MixedBatchWriter_wide_batch10k_bytes100M-10         	       8	 142323412 ns/op	    185406 bytes/op
BenchmarkWriterMemory/MixedBatchWriter_wide_defaults
BenchmarkWriterMemory/MixedBatchWriter_wide_defaults-10                   	       8	 127524651 ns/op	     96165 bytes/op
BenchmarkWriterMemory/StreamingBatchWriter_bytes100M
BenchmarkWriterMemory/StreamingBatchWriter_bytes100M-10                   	     400	   2812296 ns/op	      1393 bytes/op
BenchmarkWriterMemory/StreamingBatchWriter_defaults
BenchmarkWriterMemory/StreamingBatchWriter_defaults-10                    	     394	   2803509 ns/op	      1393 bytes/op
BenchmarkWriterMemory/StreamingBatchWriter_wide_bytes100M
BenchmarkWriterMemory/StreamingBatchWriter_wide_bytes100M-10              	       9	 120658421 ns/op	     75653 bytes/op
BenchmarkWriterMemory/StreamingBatchWriter_wide_defaults
BenchmarkWriterMemory/StreamingBatchWriter_wide_defaults-10               	       9	 117394778 ns/op	     78724 bytes/op
PASS

Process finished with the exit code 0
```

### CODE REVIEW

File: writers/writers_test.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+		bldr := array.NewRecordBuilder(memory.DefaultAllocator, sc)
+		bldr.Field(0).(*array.StringBuilder).Append("test")
+		for i := 0; i < numWideCols; i++ {
+			bldr.Field(i + 1).(*array.Int64Builder).Append(randVals[i])
+		}
+		return bldr.NewRecord()
+	}
+}
+
+func writerMatrix[T writers.Writer, C any, O ~func(T)](prefix string, constructor func(C, ...O) (T, error), client C, recordMaker func() func() arrow.Record, optsMatrix map[string][]O) []bCase {
```
Comment (disq) [inline writers/writers_test.go:145]: I think I have a problem...

File: writers/writers_test.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+		"defaults":           nil,
+		"batch10k/bytes100M": {mixedbatchwriter.WithBatchSizeBytes(100000000), mixedbatchwriter.WithBatchSize(10000)},
+	}
+	streamingbatchwriterOpts := map[string][]streamingbatchwriter.Option{
+		"defaults":  nil,
+		"bytes100M": {streamingbatchwriter.WithBatchSizeBytes(100000000)},
+	}
+
+	var bCases []bCase
+	bCases = append(bCases, writerMatrix[*batchwriter.BatchWriter, batchwriter.Client, batchwriter.Option]("BatchWriter", batchwriter.New, &batchwriterClient{}, makeRecord, batchwriterOpts)...)
```
Comment (disq) [inline writers/writers_test.go:44]: inb4, doing this:
```go
	bCases = append(bCases, writerMatrix("BatchWriter", batchwriter.New, &batchwriterClient{}, makeRecord, batchwriterOpts)...)
```

results in:
> ./writers_test.go:45:71: type *batchwriterClient of &batchwriterClient{} does not match inferred type batchwriter.Client for C

File: writers/writers_test.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+			b.StopTimer()
+
+			if err != nil {
+				b.Fatal(err)
+			}
+
+			runtime.ReadMemStats(&mEnd)
+
+			allocatedBytes := mEnd.Alloc - mStart.Alloc
+			b.ReportMetric(float64(allocatedBytes)/float64(newN), "bytes/op")
```
Comment (hermanschaaf) [inline writers/writers_test.go:89]: Doesn't Go measure memory usage already if we provide `-benchmem`? What am I missing?

Comment (disq) [reply writers/writers_test.go:89]: `benchmem` will report `allocs/op` (probably the same number as ours but I didn't compare them) and won't report until invoked with `-benchmem` or toggled from within the test. We also multiply the `N` to get more stable benchmarks (and hit writer batching spots).

**Review by hermanschaaf — APPROVED** [review]
Bit complicated logic but since it's a benchmark it's probably okay as long as it helps you find memory-related improvements 😅

File: writers/writers_test.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+		})
+		randVals[i] = rand.Int63()
+	}
+	sc := table.ToArrowSchema()
+
+	return func() arrow.Record {
+		bldr := array.NewRecordBuilder(memory.DefaultAllocator, sc)
+		bldr.Field(0).(*array.StringBuilder).Append("test")
+		for i := 0; i < numWideCols; i++ {
+			bldr.Field(i + 1).(*array.Int64Builder).Append(randVals[i])
```
Comment (candiduslynx) [inline writers/writers_test.go:139]:
```suggestion
		for i := 1; i <= numWideCols; i++ {
			bldr.Field(i).(*array.Int64Builder).Append(randVals[i])
```

File: writers/writers_test.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+			{
+				Name: "col1",
+				Type: arrow.BinaryTypes.String,
+			},
+		},
+	}
+
+	const numWideCols = 200
+	randVals := make([]int64, numWideCols)
+	for i := 0; i < numWideCols; i++ {
```
Comment (candiduslynx) [inline writers/writers_test.go:126]:
```suggestion
	randVals := make([]int64, numWideCols+1)
	for i := 1; i <= numWideCols; i++ {
```

**Review by candiduslynx — APPROVED** [review]

File: writers/writers_test.go (hunk 4 above)
Comment (disq) [reply writers/writers_test.go:136]: you'd need to do `randVals[i-1]`, adds compexity and potential copy/paste errors, just to avoid an `+ 1`

**Review by erezrokah — APPROVED** [review]

**Pull request merged**